cloudformation-states.yaml

```diff
@@ -0,0 +1,110 @@
+AWSTemplateFormatVersion: "2010-09-09"
+Description: Set up an executable pipeline state machine
+
+Parameters:
+  Environment:
+    Type: String
+    Default: staging
+    AllowedValues:
+      - staging
+      - production
+    Description: The environment for which the cluster need to be created.
+  ExperimentId:
+    Default: '6379'
+    Type: String
+    Description: The ID of the experiment the system is created for.
+
+Resources:
+  PipelineStateMachine:
+    Type: "AWS::StepFunctions::StateMachine"
+    Properties:
+      DefinitionString:
+        !Sub
+          - |-
+            {
+              "Comment": "N/A",
+              "StartAt": "DeleteCompletedJobs",
```

Comment:
mainly to learn a little bit more about this: why does the pipeline start by deleting previous jobs? couldn't this be managed somewhere else?

Reply:
It is not possible, not easily at least.

What this does is it deletes previous /completed/ jobs. Unfortunately the way Kubernetes structures Jobs is that their output persists even after they fail or succeed. This makes sense for most applications where you want to review the results after the fact.

Unfortunately for our purposes this does not work, because we want to have a Job running all the time.

Imagine if we did not delete completed jobs, and the first step is to create a Job. If the creation throws a 409 (already exists), then catch the error and fail silently, knowing that we have a job that exists and can send messages to. The problem with this is that the 409 will also be thrown if there is already a *completed* job. Therefore, there will be no Pods running and the future steps will fail.

Therefore, a necessary step for this to work is before the pipeline runs, we clean up all completed Jobs that match our specifications. This way, when the next steps throws a 409, we know that there is an *active, running* job that we can send stuff to.

There is obviously a race condition here (what if the Job shuts down in between the two steps). There is not much we can do about this and it's not the end of the world -- if no jobs are deleted in step 1, then the job finished, then in step 2 we find an already existing job and falsely assume it is running, we will fail on the first actual step because of a connection error and the state machine will fail, but the incidence of this will be quite low.